src/tests/end-to-end/tests/target-page/tabstop.test.ts

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { ElementHandle } from 'puppeteer';
+
+import { Browser } from '../../common/browser';
+import { launchBrowser } from '../../common/browser-factory';
+import { fastPassSelectors } from '../../common/element-identifiers/fastpass-selectors';
+import { tabStopShadowDomSelectors, TargetPageElementSelectors } from '../../common/element-identifiers/target-page-selectors';
+import { PopupPage } from '../../common/page-controllers/popup-page';
+import { TargetPage } from '../../common/page-controllers/target-page';
+
+describe('tabstop tests', () => {
+    let browser: Browser;
+    let targetPage: TargetPage;
+    let popupPage: PopupPage;
+
+    beforeEach(async () => {
+        browser = await launchBrowser({ suppressFirstTimeDialog: true });
+        targetPage = await browser.newTargetPage({ testResourcePath: 'native-widgets/input-type-radio.html' });
+    });
+
+    afterEach(async () => {
+        if (browser) {
+            await browser.close();
+        }
+    });
+
+    test('works when tabstop is triggered from adhoc panel', async () => {
+        popupPage = await browser.newPopupPage(targetPage);
+        await popupPage.gotoAdhocPanel();
+        await popupPage.enableToggleByAriaLabel('Tab stops');
+
+        await targetPage.bringToFront();
+        await targetPage.waitForSelector(TargetPageElementSelectors.targetPageNativeWidgetFirstRadio);
+
+        const shadowRoot = await getShadowRootAfterTabStopActivation();
+        await validateTabStopVisualizationOnTargetPage(shadowRoot);
+    });
+
+    async function getShadowRootAfterTabStopActivation(): Promise<ElementHandle<Element>> {
```

Comment:
I think this helper is combining operations that don't form a logical unit - the name of the method including a conjunction in the middle is a good sign that this is happening. I'd just inline this helper into the test body, now that there are no longer 2 different tests trying to share it.

Reply:
Ok. Makes sense. Will update the test!